### PR TITLE
[FEATURE] New class cluster for groups of similar junctions

### LINKED ISSUE #51: [FEATURE] Decouple junctions and a cluster of junctions
After we added a first simple clustering method
eldariont left a comment:
> However, I think that the implementation of the simple clustering method (including the corresponding changes to the junction class) could be improved. Proviously, the junction class represented a single junction detected from a single read. Now, it is also used to represent clusters of junctions. IMO, we should represent clusters of junctions with a decouple junctions and cluster of junctions. or a std::vector. Then, we could keep all the cluster-related information, such as the number of supporting reads, the read names, etc. separated from the information on a single junction.

We would like to create this new class.

### PR DESCRIPTION
**Pull request by eldariont**
This PR adds a new class `cluster` that represents a cluster of junctions by holding an `std::vector<junction>`. The new class implements public methods `get_average_mate1()` and `get_average_mate2` to obtain the average of all junctions in the cluster. The PR also updates the simple clustering method to use the new `cluster` class. The PR resolves #51.

I'm not very experienced with C++ and the coding guidelines yet. So please don't be shy in the reviews and point out all the potentially ugly parts of my code 🙂 Also: I was not able to run the tests on my machine (see #48) so they might be broken.

### CODE REVIEW

**Comment by Irallia** [comment]
I would pass the first review to someone else (seqan/team) and do the second review. But one point I saw directly. It is common to start structs and classes with a capital letter. Can you adjust that?

**Review by joergi-w — CHANGES_REQUESTED** [review]
The `cluster` class is not a class template, so I recommend moving the implementations into a cpp file and link them together. This can speed up compilation time a lot for your application.

**Comment by eldariont** [comment]
Thanks @joergi-w for your great suggestions. I incorporated all of them except the one below:

> The `cluster` class is not a class template, so I recommend moving the implementations into a cpp file and link them together. This can speed up compilation time a lot for your application.

I see the point of doing this but don't know the perfect way to do it. Currently, we have the executable `detect_breakends.cpp` that requires the library `detect_breakends/junction_detection.cpp` which in turn uses `cluster.hpp`. I separated `cluster.hpp` into `cluster.hpp` (declarations) and `cluster.cpp` (definitions). If I now link the library `cluster.cpp` to the executable `detect_breakends.cpp`, I get the error that `detect_breakends/junction_detection.cpp` does not know the cluster class (i.e. one library does not see the another library). What am I making wrong? Sorry for this noob question.

**Review by joergi-w — APPROVED** [review]
Thank you, looks good!
The suggestion with the cpp file is not necessary for you code to work, it was more an idea that you can still incorporate at a later time. 

To detect the cluster class, make sure that "cluster.hpp" is included and the respective library has in cmake the right `target_include_path` to find it. Then all the definitions of the header should be known. If the linking breaks (shown as: Symbol not found error), then add the cpp files that contain the missing function definitions to the respective target.

File: include/cluster.hpp
Comment (Irallia) [inline include/cluster.hpp:4]: could you sort the includes lexicographically

File: include/cluster.hpp
Comment (Irallia) [inline include/cluster.hpp:6]:
```suggestion
class Cluster
```
as we discussed, or we do this change in an extra PR where we change junctions and breakends aswell

**Review by Irallia — CHANGES_REQUESTED** [review]
LGFM, just one tiny thing 💅

File: include/cluster.hpp
Comment (eldariont) [reply include/cluster.hpp:6]: Yes, I planned to create a separate PR to keep things tidy :)

**Review by Irallia — APPROVED** [review]

**Comment by Irallia** [comment]
As a last comment: I would do this moving / splitting in cpp and hpp files in another PR (I already started with these things), as we should do this for the other classes etc aswell.

**Comment by eldariont** [comment]
> As a last comment: I would do this moving / splitting in cpp and hpp files in another PR (I already started with these things), as we should do this for the other classes etc aswell.

Yes, I agree. I got the splitting of cluster working now locally but am not sure whether it's the optimal way to do it. Maybe it would be best if you start a PR and I have a look.

**Comment by Irallia** [comment]
Ah and rebase and amend the last commit to have a tidy history before you merge ❤️

**Pull request merged**
